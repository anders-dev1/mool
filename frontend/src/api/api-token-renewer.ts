import {AccessTokenClient, RenewAccessTokenCommand } from "./api-client";
import {store} from "../redux/store";
import {sessionActions} from "../redux/actions/session.action";

export async function RenewToken() {
    let refreshToken = localStorage.getItem("refresh-token");
    if (!refreshToken){
        return Promise.reject('There was no refresh token in storage.');
    }

    // use refresh token to get new access token
    let client = new AccessTokenClient();
    let request = new RenewAccessTokenCommand();
    request.refreshToken = refreshToken;

    await client.renew(request).then(
        result => {
            localStorage.setItem("access-token", result.accessToken);
        },
        error => {
            if (error.status === 404){
                store.dispatch(sessionActions.renewTokenExpired());
            }
        }
    );
}