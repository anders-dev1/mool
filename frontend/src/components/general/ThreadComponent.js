import React from "react";
import {withTheme} from "styled-components";
import ContentContainer from "./ContentContainer";
import ProfilePictureComponent from "./ProfilePictureComponent";
import ActionPill from "./ActionPillComponent";
import TimeAgo from "react-timeago";
import {connect} from "react-redux";
import {threadActions} from "../../redux/actions/threads.action";
import {ThreadCommentsComponent} from "./ThreadCommentsComponent";

class ThreadComponent extends React.PureComponent {
    render() {
        return (
            <ContentContainer className={this.props.className}>
                <div className={"flex justify-start"}>
                    <ProfilePictureComponent className="mr1 mb2"/>
                    <div className={"flex-column pt1"}>
                        <p>{this.props.thread.author}</p>
                        {this.props.isTest === undefined &&
                            <p className={"sm"}><TimeAgo date={this.props.thread.created}/></p>}
                    </div>
                </div>
                <p className={"mb1"} data-testid={"threadContent"} >{this.props.thread.content}</p>
                <div className={"flex justify-end"}>
                    <ActionPill
                        icon={"👍"}
                        onClick={() => this.props.toggleThreadLike(this.props.thread.id)}
                        number={this.props.thread.likes}
                        color={(this.props.thread.likedByCurrentUser
                            ? this.props.theme.successColor
                            : this.props.theme.primaryColor)}
                        className={"mr1"}
                        processing={this.props.thread.isChangingLikeStatus}
                        data-testid={"threadLikeBtn"}/>
                    <ActionPill
                        icon={"💬"}
                        number={this.props.thread.numberOfComments}
                        onClick={() => this.props.toggleCommentSection(this.props.thread.id)}
                        data-testid={"threadCommentBtn"}/>
                </div>

                {this.props.thread.showingCommentSection &&
                    <ThreadCommentsComponent threadId={this.props.thread.id}/>}
            </ContentContainer>
        )
    }
}

function mapState(state, props) {
    const thread = state.threadsReducer.threadsById[props.threadId];
    return {thread}
}

const actionCreators = {
    toggleThreadLike: threadActions.toggleThreadLike,
    toggleCommentSection: threadActions.toggleCommentSection
};

const connectedComponent = withTheme(connect(mapState, actionCreators)(ThreadComponent));
export {connectedComponent as ThreadComponent};
export * from './ThreadComponent';