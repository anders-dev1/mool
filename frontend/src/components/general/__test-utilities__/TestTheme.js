export const testTheme = {
    textColor: "black",
    primaryColor: "#007BFF",
    primaryHover: "rgba(0,123,255,0.33)",
    successColor: "#28A745",
    errorColor: "#F02849",
    backgroundColor: "#F0F2F5",
    buttonTextColor: "#FFFFFF",
    dividerColor: "#C7CBD1",
    inputColor: "#99A5B0",
    spinnerColor: "#FFFFFF transparent transparent transparent",
    font: "Acumin Pro, Arial, sans-serif !important;"
};