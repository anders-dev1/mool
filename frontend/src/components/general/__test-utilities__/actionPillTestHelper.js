import {screen} from "@testing-library/react";

export function findActionPillByIcon(icon) {
    return screen.getByText(icon).closest('div');
}

export function findActionPillNumberByIcon(icon) {
    const pill = findActionPillByIcon(icon);
    const numberP = pill.lastChild.lastChild;
    return numberP.textContent;
}