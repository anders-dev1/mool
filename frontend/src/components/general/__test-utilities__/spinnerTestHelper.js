export function findSpinner(container){
    return container.getElementsByClassName('lds-ring')[0];
}