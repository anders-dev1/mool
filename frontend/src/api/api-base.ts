import fetchIntercept, {FetchInterceptorResponse} from 'fetch-intercept';
import {RenewToken} from "./api-token-renewer";

class Request {
    url: string;
    config: Config;
    constructor(url: string, config: Config) {
        this.url = url;
        this.config = config;
    }
}

const ignoredPaths: string[] = [
    "/api/accesstoken/renew",
    "/api/accesstoken/authenticate"
]

export class ApiBase {
    constructor() {
        let originalRequest: Request;
        const interceptor = fetchIntercept.register({
            request(url: string, config: Config): Promise<any[]> | any[] {
                if (ignoredPaths.some((path) => url.endsWith(path))){
                    // Calling this function removes this interceptor from the list of interceptors.
                    // This stops our interceptor from processing any longer in the current request chain.
                    interceptor();
                    return [url, config];
                }

                originalRequest = new Request(url, config);

                const token = localStorage.getItem('access-token');
                if (!token){
                    return [url, config];
                }

                const myHeaders: HeadersInit = new Headers();
                myHeaders.set('Accept', 'application/json');
                myHeaders.set('Content-Type', 'application/json');
                myHeaders.set('Authorization', "Bearer " + token);

                config.headers = myHeaders;

                return [url, config];
            },
            response(response: FetchInterceptorResponse): Promise<FetchInterceptorResponse> | any {
                if (response.status != 401) {
                    return response;
                }

                if (originalRequest.url.includes('token')) {
                    return response;
                }

                let refreshToken = localStorage.getItem("refresh-token");
                if (!refreshToken) {
                    return response;
                }

                // use refresh token to get new access token
                return RenewToken().then(
                    result => {
                        const accessToken = localStorage.getItem('access-token');

                        const myHeaders: HeadersInit = new Headers();
                        myHeaders.set('Authorization', "Bearer " + accessToken);
                        myHeaders.set('Accept', 'application/json');
                        myHeaders.set('Content-Type', 'application/json');
                        originalRequest.config.headers = myHeaders;

                        return fetch(originalRequest.url, originalRequest.config).then(
                            result => {
                                return result;
                            }
                        );
                    }
                );
            }
        });
    }
}

export interface Config {
    method: string;
    headers: HeadersInit;
}