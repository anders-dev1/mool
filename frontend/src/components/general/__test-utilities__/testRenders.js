import React from 'react'
import {render} from '@testing-library/react'
import {Provider} from 'react-redux'
import {setupStore, store} from "../../../redux/store";
import {testTheme} from "./TestTheme";
import {ThemeProvider} from "styled-components";

// This is used to set up the store for tests.
export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        testStore = setupStore(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({children}) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return {testStore, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}

// This is used with the react-test-renderer to render dom in tests.
export function internalsForRenderer(
    ui,
    {
        preloadedState = {},
        testStore = setupStore(preloadedState)
    } = {}
){
    function Wrapper(children) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return Wrapper(ui);
}