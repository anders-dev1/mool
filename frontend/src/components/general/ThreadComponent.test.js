import {CommentOrder, CommentsClient, LikeThreadClient} from "../../api";
import {internalsForRenderer, renderWithProviders} from "./__test-utilities__/testRenders";
import {ThreadComponent} from "./ThreadComponent";
import renderer from "react-test-renderer";
import {fireEvent, screen} from "@testing-library/react";
import {testTheme} from "./__test-utilities__/TestTheme";
import {findActionPillByIcon, findActionPillNumberByIcon} from "./__test-utilities__/actionPillTestHelper";
import {findSpinner} from "./__test-utilities__/spinnerTestHelper";

const testThreadId = 1;
const testThreadAuthor = "Thread Threadson";
const testThreadContent = "The thread";
const testCommentContent = "The comment";

function createState({
                         threadLiked = false,
                         likeStatusChanging = false,
                         commentsToggledOn = false,
                         numberOfComments = 1
                     } = {}) {
    return {
        threadsReducer: {
            threadsById: {
                1: {
                    id: testThreadId,
                    author: testThreadAuthor,
                    content: testThreadContent,
                    likes: 0,
                    likedByCurrentUser: threadLiked,
                    numberOfComments: numberOfComments,
                    commentOrder: CommentOrder.OldestFirst,
                    isChangingLikeStatus: likeStatusChanging,
                    showingCommentSection: commentsToggledOn,
                    fetchingComments: false,
                    postingComment: false
                }
            }
        },
        commentsReducer: {
            commentsById: {
                1: {
                    id: 1,
                    authorName: "Test Testerson",
                    content: testCommentContent,
                    likes: 0,
                    likedByCurrentUser: false,
                    isChangingLikeStatus: false
                }
            },
            threadIdToCommentIds: {
                1: [1]
            }
        }
    }
}

function renderComponent(state = createState()) {
    renderWithProviders(
        <ThreadComponent isTest threadId={testThreadId}/>,
        {preloadedState: state});
}

test("renders correctly", () => {
    const internalDom = internalsForRenderer(
        <ThreadComponent isTest threadId={testThreadId}/>,
        {preloadedState: createState()});
    const tree = renderer.create(internalDom, {}).toJSON();
    expect(tree).toMatchSnapshot();
});

test("should show name of author", () => {
    // Arrange
    renderComponent();

    // Assert
    expect(screen.getByText(testThreadAuthor)).toBeVisible();
});

test("should show the content of the thread", () => {
    // Arrange
    renderComponent();

    // Assert
    expect(screen.getByText(testThreadContent)).toBeVisible();
});

describe("like action pill click,", () => {
    describe("when tread is not liked,", () => {
        const responseDelay = 100;
        let likeCallMock;
        beforeEach(() => {
            likeCallMock = jest.spyOn(LikeThreadClient.prototype, 'like')
                .mockImplementation(() => {
                    // We need this delay for the loading spinner test.
                    // Otherwise, the loading spinner will never show.
                    return new Promise((resolve) => setTimeout(resolve, responseDelay))
                });
        });

        afterEach(() => {
            likeCallMock.mockRestore();
        });

        test("should change action pill to liked color", async () => {
            // Arrange
            const state = createState();
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);
            await new Promise((resolve) => setTimeout(resolve, responseDelay));

            // Assert
            expect(findActionPillByIcon("👍")).toHaveStyle(`background-color: ${testTheme.successColor};`);
        });

        test("should call backend like endpoint", async () => {
            // Arrange
            const state = createState();
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);
            await new Promise(process.nextTick);

            // Assert
            expect(likeCallMock).toHaveBeenCalledWith(testThreadId);
        });

        test("should display spinner while sending request", () => {
            // Arrange
            const state = createState();
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);

            // Assert
            const actionPill = findActionPillByIcon("👍");
            const spinner = findSpinner(actionPill);
            expect(spinner).toBeVisible();
        });
    });

    describe("when tread is liked,", () => {
        const responseDelay = 100;
        let unlikeCallMock;
        beforeEach(() => {
            unlikeCallMock = jest.spyOn(LikeThreadClient.prototype, 'unlike')
                .mockImplementation(() => {
                    // We need this delay for the loading spinner test.
                    // Otherwise, the loading spinner will never show.
                    return new Promise((resolve) => setTimeout(resolve, 100))
                });
        });

        afterEach(() => {
            unlikeCallMock.mockRestore();
        });

        test("should change action pill to unliked color", async () => {
            // Arrange
            const state = createState({threadLiked: true});
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);
            await new Promise((resolve) => setTimeout(resolve, responseDelay));

            // Assert
            expect(findActionPillByIcon("👍")).toHaveStyle(`background-color: ${testTheme.primaryColor};`);
        });

        test("should call backend unlike endpoint", async () => {
            // Arrange
            const state = createState({threadLiked: true});
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);
            await new Promise(process.nextTick);

            // Assert
            expect(unlikeCallMock).toHaveBeenCalledWith(testThreadId);
        });

        test("should display spinner while sending request", () => {
            // Arrange
            const state = createState();
            renderComponent(state);

            // Act
            const likeButton = findActionPillByIcon("👍");
            fireEvent.click(likeButton);

            // Assert
            const actionPill = findActionPillByIcon("👍");
            const spinner = findSpinner(actionPill);
            expect(spinner).toBeVisible();
        });
    });
});

describe("comment action pill,", () => {
    test("should display count of comments in thread", () => {
        // Arrange
        const testNumberOfComments = "36";
        const state = createState({numberOfComments: testNumberOfComments})
        renderComponent(state);

        // Assert
        const actionPillNumber = findActionPillNumberByIcon("💬");
        expect(actionPillNumber).toEqual(testNumberOfComments);
    });

    test("when clicked while comments are toggled off, should get and display comments from backend", async () => {
        // Arrange
        const newCommentContent1 = "new test comment 1";
        const newCommentContent2 = "new test comment 2";
        const newComments =
            {
                comments: [
                    {
                        id: 1,
                        content: newCommentContent1
                    },
                    {
                        id: 2,
                        content: newCommentContent2
                    }
                ]
            };
        jest.spyOn(CommentsClient.prototype, 'comments')
            .mockImplementation(() => Promise.resolve(newComments));

        const state = createState({commentsToggledOn: false});
        renderComponent(state);

        // Act
        const commentPill = findActionPillByIcon("💬");
        fireEvent.click(commentPill);
        await new Promise(process.nextTick);

        // Arrange
        expect(screen.getByText(newCommentContent1)).toBeVisible();
        expect(screen.getByText(newCommentContent2)).toBeVisible();
    });

    test("when clicked while comments toggled on, should toggle the comments component off", () => {
        // Arrange
        const state = createState({commentsToggledOn: true});
        renderComponent(state);

        // Act
        const commentPill = findActionPillByIcon("💬");
        fireEvent.click(commentPill);

        // Arrange
        expect(screen.queryByText(testCommentContent)).toBeNull();
    });
});