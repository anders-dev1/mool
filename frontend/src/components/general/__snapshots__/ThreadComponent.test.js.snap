// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bczRLJ gvFepU"
>
  <div
    className="flex justify-start"
  >
    <img
      className="sc-gsnTZi eumTWv mr1 mb2"
      src="http://placekitten.com/256/256"
    />
    <div
      className="flex-column pt1"
    >
      <p>
        Thread Threadson
      </p>
    </div>
  </div>
  <p
    className="mb1"
    data-testid="threadContent"
  >
    The thread
  </p>
  <div
    className="flex justify-end"
  >
    <div
      className="sc-eCYdqJ grpftZ mr1"
      color="#007BFF"
      data-testid="threadLikeBtn"
      href="#"
      onClick={[Function]}
    >
      <p
        className="sc-jSMfEi gBFlZU"
      >
        👍
      </p>
      <div
        className="sc-gKXOVf hWrAIB"
      >
        <p
          className="sc-iBkjds cpkijM"
          data-testid="number"
        >
          0
        </p>
      </div>
    </div>
    <div
      className="sc-eCYdqJ grpftZ"
      data-testid="threadCommentBtn"
      href="#"
      onClick={[Function]}
    >
      <p
        className="sc-jSMfEi gBFlZU"
      >
        💬
      </p>
      <div
        className="sc-gKXOVf hWrAIB"
      >
        <p
          className="sc-iBkjds cpkijM"
          data-testid="number"
        >
          1
        </p>
      </div>
    </div>
  </div>
</div>
`;
